File: src/main/java/net/adventuregame/registries/Registry.java
package net.adventuregame.registries;

import net.adventuregame.resources.ResourceLocation;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class Registry<T> {
    private final Map<ResourceLocation, T> entries = new HashMap<>();

    public void register(ResourceLocation id, T entry) {
        if (entries.containsKey(id)) throw new IllegalStateException("Duplicate: " + id);
        entries.put(id, entry);
    }

    public void register(String id, T entry) {
        register(ResourceLocation.withDefaultNamespace(id), entry);
    }

    public T get(ResourceLocation id) {
        return entries.get(id);
    }

    public T get(String id) {
        return get(ResourceLocation.fromString(id));
    }

    public Collection<T> getAll() {
        return entries.values();
    }
}

File: src/main/kotlin/net/adventuregame/registries/Registry.kt
package net.adventuregame.registries

import net.adventuregame.resources.ResourceLocation

class Registry<T> {
    private val entries: MutableMap<ResourceLocation, T> = HashMap()

    fun register(id: ResourceLocation, entry: T) {
        check(!entries.containsKey(id)) { "Duplicate: $id" }
        entries.put(id, entry)
    }

    fun register(id: String, entry: T) {
        register(ResourceLocation.Companion.withDefaultNamespace(id), entry)
    }

    fun get(id: ResourceLocation): T {
        return entries[id]!!
    }

    fun get(id: String): T {
        return get(ResourceLocation.Companion.fromString(id))
    }

    val all: MutableCollection<T>
        get() = entries.values
}
